lib/vast/CodeGen/CodeGenFunction.cpp

```diff
@@ -47,23 +47,60 @@ namespace vast::cg
     }
 
     operation function_generator::emit(const clang_function *decl) {
+        decl->dump();
         auto prototype = [&] {
             if (auto symbol = visitor.symbol(decl)) {
                 if (auto op = visitor.scope.lookup_fun(symbol.value())) {
+                    auto fn        = mlir::cast< vast_function >(op);
                     // Function declaration that is not a prototype will have zero arguments
                     // and we need to fix that when we discover them
                     // This is caused by a deprecated C feature
-                    if (auto fn = mlir::dyn_cast< mlir::FunctionOpInterface >(op)) {
-                        auto fn_args   = fn.getNumArguments();
-                        auto decl_args = decl->getNumParams();
-                        VAST_CHECK(
-                            fn_args == decl_args || fn_args == 0 || decl_args == 0,
-                            "Mismatch in number of arguments between function declaration and "
-                            "vast function."
+                    auto fn_args   = fn.getNumArguments();
+                    auto decl_args = decl->getNumParams();
+                    VAST_CHECK(
+                        fn_args == decl_args || fn_args == 0 || decl_args == 0,
+                        "Mismatch in number of arguments between function declaration and "
+                        "vast function."
+                    );
+                    if (fn_args == 0 && decl_args != 0) {
+                        fn.setType(visitor.visit(decl->getFunctionType()));
+                    }
+
+                    // inline function declarations (without definition) are unknown
+                    // because clang can not provide the information for them
+                    auto linkage = fn->getAttrOfType< core::GlobalLinkageKindAttr >("linkage");
```

Comment:
this duplicates code ... make a function that takes desired linkage, i.e., core::get_function_linkage(decl) or core::GlobalLinkageKind::ExternalLinkage and sets linkage attr and visibility